File: app/src/main/java/com/example/nfctapapp/data/repository/ChatRepository.kt
package com.example.nfctapapp.data.repository

import com.aallam.openai.api.chat.ChatMessage as OpenAIChatMessage
import com.aallam.openai.api.chat.ChatRole
import com.example.nfctapapp.data.remote.OpenAIClient
import com.example.nfctapapp.data.remote.SupabaseClient
import com.example.nfctapapp.data.remote.dto.ChatMessageDto
import com.example.nfctapapp.data.remote.dto.ChatMessageInsert
import com.example.nfctapapp.data.remote.dto.ChatSessionDto
import com.example.nfctapapp.data.remote.dto.ChatSessionInsert
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

class ChatRepository(private val userId: String) {

    private val supabase = SupabaseClient.client
    private val conversationHistory = mutableListOf<OpenAIChatMessage>()

    // 세션 관련
    suspend fun createSession(title: String? = null): Result<ChatSessionDto> {
        return try {
            val session = supabase.postgrest["chat_sessions"]
                .insert(ChatSessionInsert(userId = userId, title = title)) {
                    select()
                }
                .decodeSingle<ChatSessionDto>()
            Result.success(session)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getSessions(): Result<List<ChatSessionDto>> {
        return try {
            val sessions = supabase.postgrest["chat_sessions"]
                .select {
                    filter {
                        eq("user_id", userId)
                    }
                    order("updated_at", Order.DESCENDING)
                }
                .decodeList<ChatSessionDto>()
            Result.success(sessions)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun updateSessionTitle(sessionId: String, title: String): Result<Unit> {
        return try {
            supabase.postgrest["chat_sessions"]
                .update({ set("title", title) }) {
                    filter {
                        eq("id", sessionId)
                    }
                }
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun deleteSession(sessionId: String): Result<Unit> {
        return try {
            supabase.postgrest["chat_sessions"]
                .delete {
                    filter {
                        eq("id", sessionId)
                    }
                }
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // 메시지 관련
    suspend fun getMessages(sessionId: String): Result<List<ChatMessageDto>> {
        return try {
            val messages = supabase.postgrest["chat_messages"]
                .select {
                    filter {
                        eq("session_id", sessionId)
                    }
                    order("created_at", Order.ASCENDING)
                }
                .decodeList<ChatMessageDto>()

            // 대화 히스토리 복원
            conversationHistory.clear()
            messages.forEach { msg ->
                conversationHistory.add(
                    OpenAIChatMessage(
                        role = if (msg.isFromUser) ChatRole.User else ChatRole.Assistant,
                        content = msg.content
                    )
                )
            }

            Result.success(messages)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun saveMessage(sessionId: String, content: String, isFromUser: Boolean): Result<ChatMessageDto> {
        return try {
            val message = supabase.postgrest["chat_messages"]
                .insert(ChatMessageInsert(
                    sessionId = sessionId,
                    content = content,
                    isFromUser = isFromUser
                )) {
                    select()
                }
                .decodeSingle<ChatMessageDto>()

            // 세션 업데이트 시간 갱신
            supabase.postgrest["chat_sessions"]
                .update({ set("updated_at", "now()") }) {
                    filter {
                        eq("id", sessionId)
                    }
                }

            Result.success(message)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // AI 대화 (스트리밍)
    fun sendMessageStream(sessionId: String, userMessage: String): Flow<StreamResponse> = flow {
        // 사용자 메시지 저장
        saveMessage(sessionId, userMessage, isFromUser = true)

        val fullResponse = StringBuilder()

        try {
            // 스트리밍 응답 받기
            OpenAIClient.chatStream(userMessage, conversationHistory).collect { chunk ->
                val content = chunk.choices.firstOrNull()?.delta?.content
                if (content != null) {
                    fullResponse.append(content)
                    emit(StreamResponse.Streaming(fullResponse.toString()))
                }
            }

            val response = fullResponse.toString()

            // 대화 히스토리에 추가
            conversationHistory.add(OpenAIChatMessage(role = ChatRole.User, content = userMessage))
            conversationHistory.add(OpenAIChatMessage(role = ChatRole.Assistant, content = response))

            // AI 응답 저장
            saveMessage(sessionId, response, isFromUser = false)

            emit(StreamResponse.Complete(response))
        } catch (e: Exception) {
            emit(StreamResponse.Error(e.message ?: "메시지 전송에 실패했습니다"))
        }
    }

    fun clearHistory() {
        conversationHistory.clear()
    }
}

sealed class StreamResponse {
    data class Streaming(val content: String) : StreamResponse()
    data class Complete(val content: String) : StreamResponse()
    data class Error(val message: String) : StreamResponse()
}

File: backend/app/repositories/chat_repository.py
"""
Chat repository for Supabase operations.

Migrated from frontend ChatRepository.kt to backend for security.
"""

import logging
from typing import Optional
from datetime import datetime
from app.models import (
    ChatSessionDto,
    ChatMessageDto,
    ChatSessionInsert,
    ChatMessageInsert
)
from app.repositories.supabase_client import get_supabase_client

logger = logging.getLogger(__name__)


class ChatRepository:
    """Repository for chat sessions and messages."""

    def __init__(self):
        """Initialize repository with Supabase client."""
        self.supabase = get_supabase_client()

    # ==================== Session Operations ====================

    async def create_session(self, user_id: str, title: Optional[str] = None) -> ChatSessionDto:
        """
        Create a new chat session.

        Args:
            user_id: User ID
            title: Optional session title

        Returns:
            ChatSessionDto: Created session

        Raises:
            Exception: If creation fails
        """
        try:
            data = {
                "user_id": user_id,
                "title": title
            }

            response = self.supabase.table("chat_sessions").insert(data).execute()

            if not response.data:
                raise Exception("Failed to create session")

            session_data = response.data[0]

            return ChatSessionDto(
                id=session_data.get("id"),
                userId=session_data.get("user_id"),
                title=session_data.get("title"),
                createdAt=session_data.get("created_at"),
                updatedAt=session_data.get("updated_at")
            )

        except Exception as e:
            logger.error(f"Failed to create session: {str(e)}", exc_info=True)
            raise

    async def get_sessions(self, user_id: str) -> list[ChatSessionDto]:
        """
        Get all sessions for a user.

        Args:
            user_id: User ID

        Returns:
            list[ChatSessionDto]: List of sessions ordered by updated_at desc
        """
        try:
            response = self.supabase.table("chat_sessions") \
                .select("*") \
                .eq("user_id", user_id) \
                .order("updated_at", desc=True) \
                .execute()

            sessions = []
            for session_data in response.data:
                sessions.append(ChatSessionDto(
                    id=session_data.get("id"),
                    userId=session_data.get("user_id"),
                    title=session_data.get("title"),
                    createdAt=session_data.get("created_at"),
                    updatedAt=session_data.get("updated_at")
                ))

            return sessions

        except Exception as e:
            logger.error(f"Failed to get sessions: {str(e)}", exc_info=True)
            raise

    async def update_session_title(self, session_id: str, title: str) -> None:
        """
        Update session title.

        Args:
            session_id: Session ID
            title: New title

        Raises:
            Exception: If update fails
        """
        try:
            self.supabase.table("chat_sessions") \
                .update({"title": title}) \
                .eq("id", session_id) \
                .execute()

        except Exception as e:
            logger.error(f"Failed to update session title: {str(e)}", exc_info=True)
            raise

    async def delete_session(self, session_id: str) -> None:
        """
        Delete a session (cascade deletes messages).

        Args:
            session_id: Session ID

        Raises:
            Exception: If deletion fails
        """
        try:
            self.supabase.table("chat_sessions") \
                .delete() \
                .eq("id", session_id) \
                .execute()

        except Exception as e:
            logger.error(f"Failed to delete session: {str(e)}", exc_info=True)
            raise

    # ==================== Message Operations ====================

    async def get_messages(self, session_id: str) -> list[ChatMessageDto]:
        """
        Get all messages for a session.

        Args:
            session_id: Session ID

        Returns:
            list[ChatMessageDto]: List of messages ordered by created_at asc
        """
        try:
            response = self.supabase.table("chat_messages") \
                .select("*") \
                .eq("session_id", session_id) \
                .order("created_at", desc=False) \
                .execute()

            messages = []
            for msg_data in response.data:
                messages.append(ChatMessageDto(
                    id=msg_data.get("id"),
                    sessionId=msg_data.get("session_id"),
                    content=msg_data.get("content"),
                    isFromUser=msg_data.get("is_from_user"),
                    createdAt=msg_data.get("created_at")
                ))

            return messages

        except Exception as e:
            logger.error(f"Failed to get messages: {str(e)}", exc_info=True)
            raise

    async def save_message(
        self,
        session_id: str,
        content: str,
        is_from_user: bool
    ) -> ChatMessageDto:
        """
        Save a new message.

        Args:
            session_id: Session ID
            content: Message content
            is_from_user: True if from user, False if from AI

        Returns:
            ChatMessageDto: Created message

        Raises:
            Exception: If save fails
        """
        try:
            # Save message
            data = {
                "session_id": session_id,
                "content": content,
                "is_from_user": is_from_user
            }

            response = self.supabase.table("chat_messages").insert(data).execute()

            if not response.data:
                raise Exception("Failed to save message")

            msg_data = response.data[0]

            # Update session's updated_at timestamp
            self.supabase.table("chat_sessions") \
                .update({"updated_at": "now()"}) \
                .eq("id", session_id) \
                .execute()

            return ChatMessageDto(
                id=msg_data.get("id"),
                sessionId=msg_data.get("session_id"),
                content=msg_data.get("content"),
                isFromUser=msg_data.get("is_from_user"),
                createdAt=msg_data.get("created_at")
            )

        except Exception as e:
            logger.error(f"Failed to save message: {str(e)}", exc_info=True)
            raise


# Global repository instance
chat_repository = ChatRepository()
